### PR TITLE
BUG: Bug in quantile() and median() returned wrong result for non monotonic window borders

### LINKED ISSUE #37153: BUG: rolling.median/quantile does not work with (some) BaseIndexer subclasses
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

Using the example from `computation.rst`.

```
In [1]: In [70]: use_expanding = [True, False, True, False, True]
   ...:

In [2]: In [72]: df = pd.DataFrame({'values': range(5)})
   ...:

In [3]: In [2]: from pandas.api.indexers import BaseIndexer
   ...:

In [4]: ...: class CustomIndexer(BaseIndexer):
   ...: ...:
   ...: ...:    def get_window_bounds(self, num_values, min_periods, center, closed):
   ...: ...:        start = np.empty(num_values, dtype=np.int64)
   ...: ...:        end = np.empty(num_values, dtype=np.int64)
   ...: ...:        for i in range(num_values):
   ...: ...:            if self.use_expanding[i]:
   ...: ...:                start[i] = 0
   ...: ...:                end[i] = i + 1
   ...: ...:            else:
   ...: ...:                start[i] = i
   ...: ...:                end[i] = i + self.window_size
   ...: ...:        return start, end
   ...:

In [5]: indexer = CustomIndexer(window_size=1, use_expanding=use_expanding)

# Each window
In [7]: for window in df.rolling(indexer):
   ...:     print(window)
   ...:
   values
0       0
   values
1       1
   values
0       0
1       1
2       2
   values
3       3
   values
0       0
1       1
2       2
3       3
4       4

# Should be [0, 1, 1, 3, 2]
In [8]: df.rolling(indexer).median()
Out[8]:
   values
0     0.0
1     1.0
2     1.5
3     NaN
4     3.0

# Should be [0, 1, 1, 3, 2]
In [10]: df.rolling(indexer).quantile(0.5)
Out[10]:
   values
0     0.0
1     1.0
2     1.5
3     NaN
4     3.0
```

The cython aggregations probably assume some sort of property of the windows that may not hold with the custom indexers

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #37153
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

cc @mroeschke 

Values were not added to the window again if starting point was moved back and values were not removed from the window, when ending point was moved back

Should I create functions for the deletion and addition?

### CODE REVIEW

File: doc/source/whatsnew/v1.2.0.rst [hunk 1]
```diff
@@ -462,6 +462,7 @@ Groupby/resample/rolling
 - Bug in :meth:`RollingGroupby.count` where a ``ValueError`` was raised when specifying the ``closed`` parameter (:issue:`35869`)
 - Bug in :meth:`DataFrame.groupby.rolling` returning wrong values with partial centered window (:issue:`36040`).
 - Bug in :meth:`DataFrameGroupBy.rolling` returned wrong values with timeaware window containing ``NaN``. Raises ``ValueError`` because windows are not monotonic now (:issue:`34617`)
+- Bug in :meth:`Rolling.median()` and :meth:`Rolling.quantile()` returned wrong values for ``CustomIndexer`` with non-monotonic starting or ending points for windows (:issue:`37153`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.2.0.rst:465]: * Nit: Don't need the `()` when referencing `:meth:`s
* ``CustomIndexer`` -> ``` :class:`BaseIndexer` ``` subclasses

**Comment by mroeschke** [comment]
Yeah would be great to create a function that median and quantile could share for the skiplist inserts and removal

File: doc/source/whatsnew/v1.2.0.rst (hunk 1 above)
Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst]: Done

**Comment by phofl** [comment]
Should we move the for the ``for j in range(s, e):`` blocks into the function too? Would be differing from the other parts, but otherwise the assignment of ``err`` is a bit long winded

**Comment by mroeschke** [comment]
Hmm if it makes the code more DRY then sure.

**Comment by jreback** [comment]
this is orthogonal to #36933 ?

**Comment by phofl** [comment]
We can use ``is_monotonic_start_end_bounds`` here too nevertheless. Did miss that previously. I would rename the function after #36933 is merged.

**Comment by mroeschke** [comment]
Mind merging in master? https://github.com/pandas-dev/pandas/pull/36933 was recently pulled in (but shouldn't directly interfere)

**Comment by phofl** [comment]
Thanks for the heads up. Had to rename the method, but this was expected. I think this could be merged now.

**Comment by mroeschke** [comment]
Appears that there's another merge conflict.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**